src/Network/FKOAuthDataStore.php

```diff
@@ -0,0 +1,179 @@
+<?php
+/**
+ * @file src/Protocol/FKOAuthDataStore.php
+ * OAuth server
+ * Based on oauth2-php <http://code.google.com/p/oauth2-php/>
+ *
+ */
+namespace Friendica\Network;
+
+use Friendica\App;
+use Friendica\Core\Config;
+use Friendica\Core\System;
+use Friendica\Database\DBM;
+use dba;
+
+define('REQUEST_TOKEN_DURATION', 300);
+define('ACCESS_TOKEN_DURATION', 31536000);
+
+require_once "library/OAuth1.php";
+require_once "library/oauth2-php/lib/OAuth2.inc";
+
+/**
+ * @brief OAuthDataStore class
+ */
+class FKOAuthDataStore extends OAuthDataStore
+{
+	/**
+	 * @return string
+	 */
+	private static function genToken()
+	{
+		return md5(base64_encode(pack('N6', mt_rand(), mt_rand(), mt_rand(), mt_rand(), mt_rand(), uniqid())));
+	}
+
+	/**
+	 * @param string $consumer_key key
+	 * @return mixed
+	 */
+	public static function lookup_consumer($consumer_key)
+	{
+		logger(__function__.":".$consumer_key);
+		
+		$s = dba::select('clients', array('client_id', 'pw', 'redirect_uri'), array('client_id' => $consumer_key));
+		$r = dba::inArray($r);
+
+		if (DBM::is_result($r)) {
+			return new OAuthConsumer($r[0]['client_id'], $r[0]['pw'], $r[0]['redirect_uri']);
+		}
+
+		return null;
+	}
+
+	/**
+	 * @param string $consumer   consumer
+	 * @param string $token_type type
+	 * @param string $token      token
+	 * @return mixed
+	 */
+	public static function lookup_token($consumer, $token_type, $token)
+	{
+		logger(__function__.":".$consumer.", ". $token_type.", ".$token);
+		
+		$s = dba::select('tokens', array('id', 'secret', 'scope', 'expires', 'uid'), array('client_id' => $consumer->key, 'scope' => $token_type, 'id' => $token));
+		$r = dba::inArray($s);
+
+		if (DBM::is_result($r)) {
+			$ot=new OAuthToken($r[0]['id'], $r[0]['secret']);
+			$ot->scope = $r[0]['scope'];
+			$ot->expires = $r[0]['expires'];
+			$ot->uid = $r[0]['uid'];
+			return $ot;
+		}
+
+		return null;
+	}
+
+	/**
+	 * @param string $consumer  consumer
+	 * @param string $token     token
+	 * @param string $nonce     nonce
+	 * @param string $timestamp timestamp
+	 * @return mixed
+	 */
+	public static function lookup_nonce($consumer, $token, $nonce, $timestamp)
+	{
+		$r = dba::select('tokens', ['id', 'secret'], ['client_id' => $consumer->key, 'id' => $nonce, 'expires' => $timestamp], ['limit' => 1]);
+				
+		if (DBM::is_result($r)) {
+			return new OAuthToken($r[0]['id'], $r[0]['secret']);
```

Comment:
Since you used ```"limit" => 1```, you mustn't use ```$r[0]``` but ```$r``` here.

Reply:
see https://github.com/friendica/friendica/pull/4012